api/v1beta1/grpc_disruption.go

```diff
@@ -0,0 +1,132 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+package v1beta1
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// GRPCDisruptionSpec represents a gRPC disruption
+type GRPCDisruptionSpec struct {
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	// +ddmark:validation:Minimum=1
+	// +ddmark:validation:Maximum=65535
+	Port      int                  `json:"port,omitempty"`
+	Endpoints []EndpointAlteration `json:"endpoints,omitempty"`
+}
+
+// EndpointAlteration represents an endpoint to disrupt and the corresponding error to return
+type EndpointAlteration struct {
+	TargetEndpoint string `json:"endpoint,omitempty"`
+	// +kubebuilder:validation:Enum=OK;CANCELLED;UNKNOWN;INVALID_ARGUMENT;DEADLINE_EXCEEDED;NOT_FOUND;ALREADY_EXISTS;PERMISSION_DENIED;RESOURCE_EXHAUSTED;FAILED_PRECONDITION;ABORTED;OUT_OF_RANGE;UNIMPLEMENTED;INTERNAL;UNAVAILABLE;DATA_LOSS;UNAUTHENTICATED
+	// +ddmark:validation:Enum=OK;CANCELLED;UNKNOWN;INVALID_ARGUMENT;DEADLINE_EXCEEDED;NOT_FOUND;ALREADY_EXISTS;PERMISSION_DENIED;RESOURCE_EXHAUSTED;FAILED_PRECONDITION;ABORTED;OUT_OF_RANGE;UNIMPLEMENTED;INTERNAL;UNAVAILABLE;DATA_LOSS;UNAUTHENTICATED
+	ErrorToReturn string `json:"error,omitempty"`
+	// +kubebuilder:validation:Enum={}
+	// +ddmark:validation:Enum={}
+	OverrideToReturn string `json:"override,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=100
+	// +ddmark:validation:Minimum=1
+	// +ddmark:validation:Maximum=100
+	QueryPercent int `json:"query_pct,omitempty"`
+}
+
+const (
+	ERROR    = "error"
+	OVERRIDE = "override"
+)
+
+// Validate validates that there are no missing hostnames or records for the given grpc disruption spec
+func (s GRPCDisruptionSpec) Validate() error {
+	if s.Port == 0 {
+		return errors.New("the gRPC disruption does not specify the port which the target service exposes, but port must be specified")
+	}
+
+	if len(s.Endpoints) == 0 {
+		return errors.New("the gRPC disruption was selected with no endpoints specified, but endpoints must be specified")
+	}
+
+	queryPctByEndpoint := make(map[string]int)
+
+	for _, alteration := range s.Endpoints {
+		if alteration.TargetEndpoint == "" {
+			return errors.New("some list items in gRPC disruption are missing endpoints; specify an endpoint for each item in the list")
+		}
+
+		// check that endpoint is not already configured such that the sum of mangled queryPercents total to more than 100%
+		if totalQueryPercent, ok := queryPctByEndpoint[alteration.TargetEndpoint]; ok {
+			if alteration.QueryPercent > 0 {
+				if totalQueryPercent+alteration.QueryPercent > 100 {
+					return errors.New("total queryPercent of all alterations applied to endpoint %s is over 100%; modify them to so their total is 100% or less")
+				}
+				queryPctByEndpoint[alteration.TargetEndpoint] += alteration.QueryPercent
+			}
+		} else {
+			if alteration.QueryPercent > 100 {
```

Comment:
Wouldn't we only get this message if someone specified >100% for a single alteration? The message says that the sum of alterations exceeded 100%

Reply:
Hm, it could technically be the first of 2 but it's just over 100 anyways?

Like
`200`
`50`

I'm down to be specific too

Reply:
I am pretty sure that the error message at L67 covers the situation where we have multiple alterations that total 100, and that we can only hit this error if an individual alteration is over 100, so I think this error message (L74) should be more specific just to be clearer to users

Reply:
Hm, I reread a few times. Hear me out:

**Scenario 1:** 
If the input is: `50` then `50`, we would hit the `else` block, then on the second loop hit the `if` block, discover we reached `100` and error.

**Scenario 2:**
If the input is: `100` only, then we would hit the `else` block and in that case we'd need to catch that it was over 100% because we'd never hit the `if` block again.

**Scenario 3:**
Tricky one is this one. If we have `100` and then `50` for the same endpoint, it'll get flagged in the `else` block but it's not actually a single alteration. The sum is still `150` > `100`.

I tried to group `Scenario 2` and `Scenario 3` into the same if-statement to have less lines of code. I thought about treating lists with 1 item separately to make it clearer, but since we are iterating through and discovering as we go, that would require looping through them again after the tallies are taken.

Let me know if this sounds fishy, I think it is right.

Reply:
Hmm, two objections: (1) We only error on `> 100`, not `>= 100`, so I think technically both scenarios 1 and 2 are good and shouldn't error. (2) I've realized that because we are setting a max of 100 for any given queryPercent with ddmark, we can just drop the `if` check inside the `else`, as we should expect to literally never hit it?

Reply:
(oops yeah, should have been `51` and `50`, `101`)

Okay that's true I'll remove the if-statement!